xwiki-pro-macros-api/src/main/resources/js/tabmacro.js

```diff
@@ -0,0 +1,56 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+const delay = (delayInms) => {
+  return new Promise(resolve => setTimeout(resolve, delayInms));
+};
+
+require(['xwiki-page-ready', 'jquery'], function (pageReady, $) {
+  async function processTabGroupElement(element) {
+
+    let globalConfig = JSON.parse(element.getAttribute('data-config'));
+    let elementToIterate = element.querySelectorAll("ul.nav-tabs li a");
+
+    if (globalConfig["nextAfter"] > 0 && elementToIterate.length > 0) {
+      for (let i = 0; i < 1 || globalConfig["loopCards"]; i++) {
+        for (const el of elementToIterate) {
+          let divElement = element.querySelector("#" + el.getAttribute("aria-controls"))
+          let tabConfig = JSON.parse(divElement.getAttribute('data-config'))
+          if (tabConfig["nextAfter"] <= 0) {
+            // should not happen but in case avoid to have an infinite loop
+            return
+          }
+          $(el).tab('show')
+          await delay(tabConfig["nextAfter"] * 1000)
+        }
+      }
+    }
+  }
+
+  pageReady.afterPageReady(() => {
+    let tabMacroElements = document.querySelectorAll('div.xwikitabmacro');
+    try {
+      let promises = tabMacroElements.values().map(i => processTabGroupElement(i));
+      Promise.all(promises).catch((err) => console.log(err));
+    }
+    catch (err) {
+      console.error(err);
+    }
+  });
+});
```

Comment:
This code could use some comments, it is super hard to understand, and it is not clear to me why it does things the way it does them. For example, why does it introduce a global method `delay`? Why is delay returning a promise that just calls `setTimeout`? Why is this async and not just a regular method? Also, it seems weird to me that the whole configuration is serialized as JSON if only two values out of the configuration are actually needed, it would seem much more natural to me to just store these two values as data attributes.

Reply:
> For example, why does it introduce a global method delay?

I'll inline it we don't really need it.

> Why is this async and not just a regular method ?

Because I feel that it was the easiest way to implement this. But maybe there are simpler way to do this.

> Also, it seems weird to me that the whole configuration is serialised as JSON if only two values out of the configuration are actually needed, it would seem much more natural to me to just store these two values as data attributes.

Because I thought to extends the feature of this macro it's easier to have all configuration in the js part. But yes I agree that now it's a bit overkill. I'll do what you suggested.

Reply:
Okay, I understand the idea now, I think it would be good to add a comment what this `setTimeout` does (sleep) for those who don't know the pattern. Also, this for-loop doesn't feel natural. I wonder if it wouldn't be clearer to use a [do...while](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/do...while) loop with `loopCards === "true"` as condition. That should have the same effect but is much more explicit.